### ISSUE TITLE
pageserver: concurrent IO on the get page read path

### ISSUE THREAD

**Issue by VladLazar** [neondatabase/neon#9378]
The read path does sequential IOs.

We should issue IO concurrently within one layer visit and to move on to the next layer without waiting for IOs
from the previous visit to complete.

**Comment by problame** [comment]
Status update:
- impl is in https://github.com/neondatabase/neon/pull/9353
- Vlad started, I took over
- today, I merged in https://github.com/neondatabase/neon/pull/10386
  because that PR is going to get merged and deployed first; conflicts around protocol version v3 (super high confidence the merge was correct, and @skyzh 's https://github.com/neondatabase/neon/pull/10313 )
- got rid of the barrier stage, as discussed with Vlad before the holidays
- that appears to have fixed another mysterious hang bug (?)
- removed todo() and deprecation warning ; just rename it to serial() and remove the warning?

Next steps:

- clean up `deep_layers_with_delta.py` script / make it reusable and use in batching benchmark
- clean up ad-hoc changes made to benchmarks (new params are ok, removing param space not)
- convince ourselves that the current design works fine with limited slots, because the background task advances the io futures ( => don't need https://github.com/neondatabase/tokio-epoll-uring/pull/63 )
- review interaction with sparse keyspace
  - sparked by the minor merge conflicts with https://github.com/neondatabase/neon/pull/10313
  - but really, I don't think either vlad or me thought about this at all
- review teardown / VirtualFile lifecycle 
  - basically: because the VirtualFile is now wrapped in an Arc<> that lives inside the spawned ios, thus potentially longer than the get_values_reconstruct_data call that spawned the io, can it happen that compaction deletes the file from the filesystem while the Arc<VirtualFile> still exists?
- `  // TODO: IO futures are not failable - we could expect`
  - at a minimum, understand how exactly we're behaving if we bail on the second of three `on_disk_values`
- figure out how to monitor impact in staging/prod wrt compaction debt
  - maybe just do a big ingest before and after enabling it?

**Comment by problame** [comment]
Had a call with Vlad:

> convince ourselves that the current design works fine with limited slots, because the background task advances the io futures ( => don't need https://github.com/neondatabase/tokio-epoll-uring/pull/63 )

We don't need it because each of the IO futures only uses one tokio-epoll-uring slot at a time, i.e., there's never a "hold tokio-epoll-uring slot and try to acquire another one".

However, that's totally non-obvious and thus brittle.
Should fix that _after_ we have removed the `Serial` implementation and transition away from opaque IO futures to a well-defined operation that's like "do _one_ vectored read against this VirtualFile and collect the results into this oneshot".

> review interaction with sparse keyspace

We convinced ourselves that conurrent IO and sparse keyspace are orthogonal.

The intersection between the two is only in the traversal code around `update_key`.
https://github.com/neondatabase/neon/blob/8c5005ff59d1556435bf47d7d9b6610f770666ba/pageserver/src/tenant/storage_layer.rs#L415-L456

Before, it was called after IO completed and used the `Value` to decide whether to mark state as complete + addd do keys_done yes/no .

Now, instead of a read Value, we have argument `completes`. 

Looking at the code, it's equivalent.


> review teardown / VirtualFile lifecycle

The current code extends the lifetime of the "raw" file abstractions too much, i.e.
* Delta and Image layer's VirtualFile is extended beyond the lifetime of the DownloadedLayer
  * This can cause eviction to happen, local file to be removed, and thus spawned IO to fail.
  * This is a bug and needs to be fixed before merging.
* InMemoryLayer's EphemeralFile lifetime is extended
  * No bug potential because the `EphemeralFile::drop` is what does the deletion from disk.

Decision: hold `DownloadedLayer` and `InMemoryLayer` alive in the spawned IO futures, instead of the "raw" file abstractions VirtualFile / EphemeralFile. That fixes the bug AND expresses more clearly to future readers of the code the implications of the spawned io futures. 

>   // TODO: IO futures are not failable - we could expect

Didn't get to this. Will DM with Vlad about it.

---

The other issues will be addressed post-merge when we experiment in staging:

* new automated benchmarks utilizing L0 layer stack
* slots performance implications
* figure out how to monitor impact in staging/prod wrt compaction debt

**Comment by problame** [comment]
> // TODO: IO futures are not failable - we could expect


The TODO was just an observation while writing the code.

Regarding error handling in general, what we want as part of the first cut of this PR is that when get_vectored_impl returns, even with an error, all IOs are done.
For now we can assume get_vectored_impl is polled to completion.

That's not the case right now, we'll return a pagestream error and move on to the next message, while the IOs from the failure are still executing inside the sidecard task.

Ideas (I will think of something):
- posion pill message to cancel all preceding IOs?
- cancellation token / dropguard in the ValuesReconstructData?
